File: scala/collatz-conjecture/src/main/scala/CollatzConjecture.scala
import scala.annotation.tailrec

object CollatzConjecture {
  @tailrec
  def step(n: Int, acc: Int): Int = n match {
    case 1               => acc
    case x if x % 2 == 0 => step(x / 2, acc + 1)
    case _               => step(3 * n + 1, acc + 1)
  }

  def steps(n: Int): Option[Int] =
    if (n <= 0) None
    else Some(step(n, 0))
}

File: fsharp/collatz-conjecture/CollatzConjecture.fs
module CollatzConjecture

let steps (number: int) : int option =
    let rec step number acc =
        match number with
        | 1                -> acc
        | n when n % 2 = 0 -> step (n / 2) (acc + 1)
        | _                -> step (3 * number + 1) (acc + 1)

    if number <= 0 then None
    else Some(step number 0)
